tools/cmake/cbmc_recipes.cmake

```diff
@@ -0,0 +1,323 @@
+nonempty list. You should only do one or the other. Value of
+cbmc_nondet_static_exclude: [${cbmc_nondet_static_exclude}].")
+endif()
+
+set(cbmc_proof_names "${cbmc_proof_name};${cbmc_proof_names}" CACHE INTERNAL "")
+list(APPEND cbmc_test_labels ${cbmc_proof_name})
+
+
+# ______________________________________________________________________________
+#     Common recipes for building a single proof
+# ``````````````````````````````````````````````````````````````````````````````
+
+# Separately build an object file from the proof harness, and an object file
+# from all project source files that the harness will link against.  This is
+# because we sometimes want to remove particular functions out of the project,
+# and particular functions out of the harness. So we remove the functions from
+# each object file and then link them together.
+#
+# After the project and harness have been linked together, we perform a series
+# of transformations on the resulting object file using goto-instrument. The
+# result of each transformation is the input to the next one. We leave behind
+# all intermediate files, with sequence numbers, to aid debugging.
+
+list(APPEND cbmc_compile_definitions
+    CBMC_OBJECT_BITS=${cbmc_object_bits}
+    CBMC_MAX_OBJECT_SIZE=${cbmc_max_object_size}
+)
+
+add_executable(
+    ${cbmc_proof_name}_0010_harness.goto
+    ${cbmc_harness_source}
+)
+
+set_target_properties(
+    ${cbmc_proof_name}_0010_harness.goto
+    PROPERTIES
+    INCLUDE_DIRECTORIES "${cbmc_compile_includes}"
+    COMPILE_DEFINITIONS "${cbmc_compile_definitions}"
+    COMPILE_OPTIONS     "${cbmc_compile_options}"
+)
+
+add_executable(
+    ${cbmc_proof_name}_0010_project.goto
+    ${cbmc_project_source}
+)
+
+set_target_properties(
+    ${cbmc_proof_name}_0010_project.goto
+    PROPERTIES
+    INCLUDE_DIRECTORIES "${cbmc_compile_includes}"
+    COMPILE_DEFINITIONS "${cbmc_compile_definitions}"
+    COMPILE_OPTIONS     "${cbmc_compile_options}"
+)
+
+# Function removal
+#
+# If funs_to_remove is a non-empty list of functions, return a goto-instrument
+# command that strips those functions out of the binary.  If the list is empty,
+# return a command to copy the binary without changing it.
+function(cbmc_remove_functions funs_to_remove binary)
+    list(LENGTH ${funs_to_remove} n_funs_to_remove)
+
+    if(${n_funs_to_remove})
+      list(JOIN ${funs_to_remove} ";--remove-function-body;" cmd_list)
+      list(PREPEND
+          cmd_list
+          goto-instrument
+          --verbosity ${CBMC_VERBOSITY}
+          --remove-function-body
+      )
+      list(APPEND
+          cmd_list
+          ${cbmc_proof_name}_0010_${binary}.goto
+          ${cbmc_proof_name}_0020_${binary}_functions_removed.goto
+      )
+
+    else()
+      set(cmd_list
+          cmake -E copy
+          ${cbmc_proof_name}_0010_${binary}.goto
+          ${cbmc_proof_name}_0020_${binary}_functions_removed.goto
+      )
+    endif()
+
+    add_custom_command(
```

Comment:
I recommend moving this custom command definition inside the if-then-else.  Then the command can be easily stated (just reference the interpolated list), and the comment can be set to something sensible.